pkg/agent/client/endpoint_resolver.go

```diff
@@ -0,0 +1,292 @@
+		return
+	}
+
+	// We only start one worker for this controller.
+	go wait.Until(r.runWorker, time.Second, ctx.Done())
+
+	<-ctx.Done()
+}
+
+func (r *EndpointResolver) runWorker() {
+	for r.processNextWorkItem() {
+	}
+}
+
+func (r *EndpointResolver) processNextWorkItem() bool {
+	key, quit := r.queue.Get()
+	if quit {
+		return false
+	}
+	defer r.queue.Done(key)
+
+	if err := r.resolveEndpoint(); err == nil {
+		r.queue.Forget(key)
+	} else {
+		klog.ErrorS(err, "Failed to resolve Service Endpoint, requeuing", "key", key)
+		r.queue.AddRateLimited(key)
+	}
+
+	return true
+}
+
+func (r *EndpointResolver) resolveEndpoint() error {
+	klog.V(2).InfoS("Resolving Endpoint", "service", klog.KRef(r.namespace, r.serviceName))
+	endpointURL, err := proxy.ResolveEndpoint(r.serviceLister, r.endpointsLister, r.namespace, r.serviceName, r.servicePort)
+	// Typically we will get one of these 2 errors (unavailable or not found).
+	// In this case, it makes sense to reset the Endpoint URL to nil and notify listeners.
+	// There is also no need to retry, as we won't find a suitable Endpoint until the Service or
+	// the Endpoints resource is updated in a way that will cause this function to be called again.
+	if errors.IsServiceUnavailable(err) {
+		klog.ErrorS(err, "Cannot resolve endpoint because Service is unavailable", "service", klog.KRef(r.namespace, r.serviceName))
+		r.updateEndpointIfNeeded(nil)
+		return nil
+	}
+	if errors.IsNotFound(err) {
+		klog.ErrorS(err, "Cannot resolve endpoint because of missing resource", "service", klog.KRef(r.namespace, r.serviceName))
+		r.updateEndpointIfNeeded(nil)
+		return nil
+	}
+	if err != nil {
+		// Unknown error: we err on the side of caution.
+		// Do not reset the URL or notify listeners, and trigger a retry.
+		return err
+	}
+	klog.V(2).InfoS("Resolved Endpoint", "service", klog.KRef(r.namespace, r.serviceName), "url", endpointURL)
+	r.updateEndpointIfNeeded(endpointURL)
+	return nil
+}
+
+func (r *EndpointResolver) updateEndpointIfNeeded(endpointURL *url.URL) {
+	// The separate Load and Store calls are safe because there is a single writer for r.endpointURL.
+	currentEndpointURL := r.endpointURL.Load()
+	updateNeeded := func() bool {
+		if endpointURL == nil && currentEndpointURL == nil {
+			return false
+		}
+		if endpointURL == nil || currentEndpointURL == nil {
+			return true
+		}
+		return endpointURL.String() != currentEndpointURL.String()
+	}
+	if !updateNeeded() {
+		klog.V(2).InfoS("No change to Endpoint for Service, no need to notify listeners", "service", klog.KRef(r.namespace, r.serviceName))
+		return
+	}
+	if endpointURL != nil {
+		klog.InfoS("Selected Endpoint has changed for Service, notifying listeners", "service", klog.KRef(r.namespace, r.serviceName), "url", endpointURL)
+	} else {
+		klog.InfoS("No more selected Endpoint for Service, notifying listeners", "service", klog.KRef(r.namespace, r.serviceName))
```

Comment:
nit: How about the following logs to make it a bit more obvious they are two results of the same operation:
- "Selected a new Endpoint for Service, notifying listeners"
- "Selected no Endpoint for Service, notifying listeners"